### PR TITLE
Add plugin upload and playground setup to PR builds

### PR DESCRIPTION
**Pull request by yscik**
### Changes proposed in this Pull Request

* Change the plugin build step action to upload the plugin zip file to wpjobmanager.com's media library, instead of as an artifact of the action
* Post the plugin zip link and a link to open it in a testing site ~as a comment~ to the PR
* Delete previous zips when a PR is updated or merged

See p6r3EZ-1PL-p2 for more details. 

### Testing instructions

* See GH action logs for this PR: https://github.com/Automattic/WP-Job-Manager/actions/runs/6176999314/job/16767164534?pr=2577
* See generated comment below

----

| Plugin build for d7b3e6dafe75c6d63749a71b7df2f249205ebbad <a href="#"></a> |
| ------------------------------------------------------------ |
| 📦 [Download plugin zip](https://wpjobmanager.com/wp-content/uploads/2023/09/wp-job-manager-zip-2577-d7b3e6da.zip)                       |
| ▶️ [Open in playground](https://wpjobmanager.com/playground/?core=2023/09/2577-d7b3e6da)             |

### CODE REVIEW

**Comment by yscik** [comment]
Updated to put the links in the PR description instead of a comment, to prevent getting notifications each time the build is run

**Comment by fjorgemota** [comment]
FYI, one approach that IMHO works well is to just update the comment previously posted instead of posting a new comment each time.

I used that approach with WordPress Dependencies Report (p6rkRX-5Aw-p2), and it works especially well because it also avoids an additional notification each time the build is run. :eyes: I think Jetpack also does something similar to this, btw.

**Comment by yscik** [comment]
There is still an initial notification for the PR author for the first comment which I didn't like :) It seems right to put it close to 'Testing instructions', but we can change it to whatever is everyone's preference.

File: scripts/upload-and-link-plugin-zip.mjs [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+
+	const oldZips = JSON.parse( execSync( `curl -s -u "${ login }" "${ MEDIA_LIBRARY_ENDPOINT }?mime_type=application/zip&_fields=id,date,title,source_url"` ).toString() );
+	if ( oldZips?.code || ! Array.isArray( oldZips ) ) {
+		console.log( `[${ oldZips.code }]`, oldZips.message );
+		return;
+	}
+	oldZips?.forEach( ( zip ) => {
+		const
+			title = zip.title.rendered;
+		if ( title.match( new RegExp( `wp-job-manager-zip-${pr}` ) ) ) {
```
Comment (fjorgemota) [inline scripts/upload-and-link-plugin-zip.mjs]: Can't this be replaced by a simpler `.startsWith`?

```suggestion
		if ( title.startsWith( `wp-job-manager-zip-${pr}-`  ) ) {
```

File: scripts/upload-and-link-plugin-zip.mjs [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+	console.log( chalk.green( '✓' ), `Plugin file uploaded to ${ zip }` )
+
+	return zip;
+}
+
+/**
+ * Post a comment on the PR with a link to the plugin zip and playground.
+ * @param {string} zip URL to plugin zip file.
+ */
+function addComment( zip ) {
```
Comment (fjorgemota) [inline scripts/upload-and-link-plugin-zip.mjs]: Probably we will want to update the function name/jsdoc to match the new intent?

File: scripts/upload-and-link-plugin-zip.mjs [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+( function main() {
+
+	// deleteOldZips();
+	//
+	// if ( merged ) {
+	// 	return;
+	// }
+	//
+	// const zip = uploadZip();
+	const zip = "https://wpjobmanager.com/wp-content/uploads/2023/09/wp-job-manager-zip-2577-26255dc9.zip";
```
Comment (fjorgemota) [inline scripts/upload-and-link-plugin-zip.mjs]: Did you change this just for the sake of the demo?

File: scripts/upload-and-link-plugin-zip.mjs [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+/**
+ * Upload plugin zip to media library.
+ *
+ * @returns {string} URL to plugin zip file.
+ */
+function uploadZip() {
+
+	const id = `${ pr }-${ commit.substring( 0, 8 ) }`;
+
+	const zip = execSync( `curl -u "${ login }" --http1.1 --data-binary @wp-job-manager.zip -H "Content-Disposition: attachment; filename=\"wp-job-manager-zip-${ id }.zip\"" ${ MEDIA_LIBRARY_ENDPOINT }?title=wp-job-manager-zip-${ id } | jq .source_url` ).toString().replaceAll( '"', '' ).trim();
```
Comment (fjorgemota) [inline scripts/upload-and-link-plugin-zip.mjs]: If a non-automattician forks this repo, and then send a PR to it, then `login` will be empty, as [the secret isn't available in these scenarios](https://docs.github.com/en/actions/managing-workflow-runs/approving-workflow-runs-from-public-forks). Maybe we shold avoid trying to run the `curl` command if `login` is empty? (and then avoid modifying the description too, naturally?)

Also, what happens if this command fails to run properly?

File: scripts/upload-and-link-plugin-zip.mjs [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+function addComment( zip ) {
+
+	const [ , path, id ] = zip.match( 'wp-content/uploads/(.*)/wp-job-manager-zip-(.*).zip' );
+	const playgroundLink = `https://wpjobmanager.com/playground/?core=${ path }/${ id }`
+
+	const links = `
+<!-- wpjm:plugin-zip -->
+----
+
+| Plugin build for ${ commit } <a href="#"><img width=600></a> |
```
Comment (fjorgemota) [inline scripts/upload-and-link-plugin-zip.mjs:106]: Do we need this link? 

```suggestion
| Plugin build for ${ commit } |
```

**Review by fjorgemota — COMMENTED** [review]
Hey there, this is looking generally good.

IMHO, I feel like having a separate comment is better than just changing the description, especially because the description is something that the user can edit to point anywhere else, but a comment from `github-actions` cannot.

With that said, I found a few issues that I think are worth mentioning. =)

Thanks for the PR!

File: scripts/upload-and-link-plugin-zip.mjs (hunk 3 above)
Comment (yscik) [reply scripts/upload-and-link-plugin-zip.mjs]: 🤦  Yup, restored in 138a689

File: scripts/upload-and-link-plugin-zip.mjs (hunk 4 above)
Comment (yscik) [reply scripts/upload-and-link-plugin-zip.mjs]: Good catch, I added an early exit. Generally if anything throws an error it should make it to the workflow's logs, but made it a bit more graceful ([`commits`](https://github.com/Automattic/WP-Job-Manager/pull/2577/files/64e63fba8aa437c5e6be6606af534ea97df47282..265b709cac8676e43712f1818d30f552b8bd8876))

File: scripts/upload-and-link-plugin-zip.mjs (hunk 5 above)
Comment (yscik) [reply scripts/upload-and-link-plugin-zip.mjs:106]: The image is a hack for making the table full-width, and the link is a hack to override the default link generated for images 👨‍💻

Comment (fjorgemota) [reply scripts/upload-and-link-plugin-zip.mjs:106]: Oh, that's neat. Nice! Thanks for explaining what it is. :)

File: scripts/upload-and-link-plugin-zip.mjs (hunk 2 above)
Comment (yscik) [reply scripts/upload-and-link-plugin-zip.mjs]: Updated in [859fe9e](https://github.com/Automattic/WP-Job-Manager/pull/2577/commits/859fe9e9cc62677e10113374680768ae18004cb9)

File: .github/workflows/build.yml [hunk 6]
```diff
@@ -1,49 +1,54 @@
...
+        run: npm run build
+      - name: Decompress plugin
+        run: unzip wp-job-manager.zip -d wp-job-manager
+      - name: Store Artifact
+        uses: actions/upload-artifact@v2
+        with:
+          name: wp-job-manager-${{ github.event.pull_request.head.sha }}
+          path: ${{ github.workspace }}/wp-job-manager/
+          retention-days: 7
+      - name: Save plugin zip and add link to it in a comment
```
Comment (fjorgemota) [inline .github/workflows/build.yml]: Just a reminder to update this:

```suggestion
      - name: Save plugin zip and add link to the PR description
```

**Review by fjorgemota — APPROVED** [review]
Looks good and works well!

I see two possible improvements for this task. One being a must, another being more optional:

1. I created an issue on WPJMCOM to handle the first issue: 654-gh-Automattic/wpjobmanager.com, just to avoid talking about the issue publicly
2. The second is moving most of the logic to WPJMCOM (so e.g. it downloads files from Github instead of uploading files from Github to WPJMCOM, as well as handle the ZIP files appropriately). This isn't a requirement to make this work.

With that said, I'm already approving the issue. I also left a suggestion to fix the name of the action in the `build.yml` workflow, too.

Thanks!

**Pull request merged**